backend/database.js

```diff
@@ -52,57 +52,91 @@ export default ((sbp('sbp/selectors/register', {
     }
     // Number of entries pushed.
     let counter = 0
-    let { hash: currentHash, ...serverMeta } = (
-      await sbp('chelonia/db/getEntryMeta', contractID, height)
-    ) || {}
+    let currentHeight = height
+    let currentHash, serverMeta
     let prefix = ''
-    let ended = false
-    // NOTE: if this ever stops working you can also try Readable.from():
-    // https://nodejs.org/api/stream.html#stream_stream_readable_from_iterable_options
-    const stream = new Readable({
-      construct (cb) {
-        this.push('[')
-        cb()
-      },
-      read (): void {
-        if (ended) {
-          this.destroy()
-          return
+    const nextKeyOp = (() => {
+      let index
+
+      return async () => {
+        if (!index) {
+          index = (await sbp('chelonia.db/get', `_private_keyop_idx_${contractID}_${currentHeight - currentHeight % 10000}`))?.split('\x00')
         }
-        if (currentHash && counter < limit) {
-          sbp('chelonia/db/getEntry', currentHash).then(async entry => {
-            if (entry) {
-              const currentPrefix = prefix
-              prefix = ','
-              counter++
-              const { hash: newCurrentHash, ...newServerMeta } = (
-                await sbp('chelonia/db/getEntryMeta', contractID, entry.height() + 1)
-              ) || {}
-              this.push(
-                `${currentPrefix}"${strToB64(
-                  JSON.stringify({ serverMeta, message: entry.serialize() })
-                )}"`
-              )
-              currentHash = newCurrentHash
-              serverMeta = newServerMeta
-            } else {
-              this.push(']')
-              this.push(null)
-              ended = true
-            }
-          }).catch(e => {
-            console.error(`[backend] streamEntriesAfter: read(): ${e.message}:`, e.stack)
-            this.push(']')
-            this.push(null)
-            ended = true
-          })
+        const value = index?.find((h, i) => {
+          if (Number(h) >= currentHeight) {
+            // Remove values that no longer are relevant from the index
+            index = ((index: any): any[]).splice(i + 1)
```

Comment:
Since using `splice()` mutates the original array, making this a bit hard to reason about.

Could you use `slice` instead?

```diff
-          index = ((index: any): any[]).splice(i + 1)
+          index = (index: any[]).slice(i + 1)
```

(Suggestion by `o4-mini-high` I agree with)

Reply:
Splice is significantly faster because it doesn't need to make a copy, and in this case the entire point is discarding data, and there's no danger of mutating an array reference.

https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJ3ajRPM1ZuTjIwLV9NUzRUR3pFODIiLCJjb2RlIjoiREFUQS5zcGxpY2UoNTAwKSIsIm5hbWUiOiJTcGxpY2UiLCJkZXBlbmRlbmNpZXMiOltdfSx7ImlkIjoiMEFKa0QycFJ6WG8xSFdGc0dQeWhMIiwiY29kZSI6IkRBVEEuc2xpY2UoNTAwKSIsIm5hbWUiOiJTbGljZSIsImRlcGVuZGVuY2llcyI6W119XSwiY29uZmlnIjp7Im5hbWUiOiJCYXNpYyBleGFtcGxlIiwicGFyYWxsZWwiOnRydWUsImdsb2JhbFRlc3RDb25maWciOnsiZGVwZW5kZW5jaWVzIjpbXX0sImRhdGFDb2RlIjoicmV0dXJuIG5ldyBBcnJheSgxMDI0KS5maWxsKDUpIn19

Reply:
Never mind, the above benchmark isn't entirely accurate because `splice` will mutate the array and `DATA` isn't reset on each run. See the more accurate benchmark below.

Since they seem to have similar performance (and `slice` may actually be faster), I'll make this change.


https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJ3ajRPM1ZuTjIwLV9NUzRUR3pFODIiLCJjb2RlIjoiREFUQVswXSA9IERBVEFbMF0uc3BsaWNlKDUwMCkiLCJuYW1lIjoiU3BsaWNlIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6IjBBSmtEMnBSelhvMUhXRnNHUHloTCIsImNvZGUiOiJEQVRBWzBdID0gREFUQVswXS5zbGljZSg1MDApIiwibmFtZSI6IlNsaWNlIiwiZGVwZW5kZW5jaWVzIjpbXX1dLCJjb25maWciOnsibmFtZSI6IkJhc2ljIGV4YW1wbGUiLCJwYXJhbGxlbCI6dHJ1ZSwiZ2xvYmFsVGVzdENvbmZpZyI6eyJkZXBlbmRlbmNpZXMiOltdfSwiZGF0YUNvZGUiOiJyZXR1cm4gW25ldyBBcnJheSgxMDI0KS5maWxsKDUpXSJ9fQ